ggml-cuda.cu

```diff
@@ -726,115 +645,335 @@ static void ggml_cuda_mul_mat_f16(const ggml_tensor * src0, const ggml_tensor *
-    char  * d_Q = (char  *) ggml_cuda_pool_malloc(n_mm * q_sz, &q_size);
+    const size_t src0_ts = ggml_type_size(src0->type);
+    const size_t src0_bs = ggml_blck_size(src0->type);
 
-    const to_fp32_cuda_t to_fp32_cuda = ggml_get_to_fp32_cuda(type);
-    dequantize_mul_mat_vec_cuda_t dmmv = ggml_get_dequantize_mul_mat_vec_cuda(type);
-    GGML_ASSERT(to_fp32_cuda != nullptr);
+    const bool src0_on_device = src0->backend == GGML_BACKEND_GPU || src0->backend == GGML_BACKEND_GPU_SPLIT;
+    const bool src0_is_f32 = src0->type == GGML_TYPE_F32;
+    const bool src0_needs_f32 = op_type & 0x4; // 3rd least significant bit = src0 needs f32
 
-    for (int64_t i03 = 0; i03 < ne03; i03++) {
-        for (int64_t i02 = 0; i02 < ne02; i02++) {
-            int i = i03*ne02 + i02;
-            cudaStream_t cudaStream = g_cudaStreams[i % GGML_CUDA_MAX_STREAMS];
-            cudaStream_t cudaStream2 = g_cudaStreams2[i % GGML_CUDA_MAX_STREAMS];
-            cudaEvent_t  cudaEvent = g_cudaEvents[i % GGML_CUDA_MAX_EVENTS];
+    const bool src1_on_device = src1->backend == GGML_BACKEND_GPU || src1->backend == GGML_BACKEND_GPU_SPLIT;
 
-            float * c_Y = d_Y + i * y_ne;
-            float * c_D = d_D + i * d_ne;
-            char  * c_Q = d_Q + i * q_sz;
+    const bool dst_on_device = dst->backend == GGML_BACKEND_GPU || dst->backend == GGML_BACKEND_GPU_SPLIT;
 
-            // copy src0 to device if necessary
-            if (src0->backend == GGML_BACKEND_CPU) {
-                CUDA_CHECK(ggml_cuda_h2d_tensor_2d(c_Q, src0, i03, i02, cudaStream2));
-            } else if (src0->backend == GGML_BACKEND_CUDA) {
-                c_Q = ((char *) src0->data) + i * q_sz;
-            } else {
-                GGML_ASSERT(false);
-            }
-            if (mul_mat_vec) { // specialized dequantize_mul_mat_vec kernel
-                CUDA_CHECK(cudaEventRecord(cudaEvent, cudaStream2));
+    const to_fp32_cuda_t to_fp32_cuda = ggml_get_to_fp32_cuda(src0->type);
 
-                // copy src1 to device
-                CUDA_CHECK(ggml_cuda_h2d_tensor_2d(c_Y, src1, i03, i02, cudaStream));
+    // dd = data device
+    char  * src0_ddq[GGML_MAX_DEVICES] = {nullptr}; // quantized
+    float * src0_ddf[GGML_MAX_DEVICES] = {nullptr}; // float
+    float * src1_ddf[GGML_MAX_DEVICES] = {nullptr};
+    float * dst_ddf[GGML_MAX_DEVICES] = {nullptr};
 
-                // wait for data
-                CUDA_CHECK(cudaStreamWaitEvent(cudaStream, cudaEvent, 0));
+    // asq = actual size quantized, asf = actual size float
+    size_t src0_asq[GGML_MAX_DEVICES];
+    memset(src0_asq, 0, sizeof(size_t)*GGML_MAX_DEVICES);
+    size_t src0_asf[GGML_MAX_DEVICES];
+    memset(src0_asf, 0, sizeof(size_t)*GGML_MAX_DEVICES);
+    size_t src1_asf[GGML_MAX_DEVICES];
+    memset(src1_asf, 0, sizeof(size_t)*GGML_MAX_DEVICES);
+    size_t dst_asf[GGML_MAX_DEVICES];
+    memset(dst_asf, 0, sizeof(size_t)*GGML_MAX_DEVICES);
```

Comment:
Minor nit, is there any reason to use `memset` here instead of initializing to `= {0}`, similar to the previous declarations?

Reply:
Would that actually initialize the whole array? I think to remember that in C the initializer `= {0}` will only initialize the first value; I only added the `= {nullptr}` for readability.

Reply:
Yes, this would initialize all the elements to zero ([ref](https://en.cppreference.com/w/c/language/array_initialization)).